### PR TITLE
Feat/markev chain 2

### PR DESCRIPTION
**Pull request by rdanitz**

### CODE REVIEW

File: quickcheck-state-machine.cabal [hunk 1]
```diff
@@ -57,13 +57,15 @@ library
         hmatrix,
         markov-chain-usage-model,
         mtl >=2.2.1,
+        optparse-applicative,
```
Comment (stevana) [inline quickcheck-state-machine.cabal]: Needed? Not reexported by Tasty?

File: src/Test/StateMachine/Markov.hs [hunk 2]
```diff
@@ -285,24 +288,54 @@ toMatrix m = do
 
 ------------------------------------------------------------------------
 
+instance IsString str => MonadFail (Either str) where
```
Comment (stevana) [inline src/Test/StateMachine/Markov.hs]: Not used? Also not worth `{-# OPTIONS_GHC -Wno-orphans #-}`?

File: src/Test/StateMachine/Markov.hs [hunk 3]
```diff
@@ -285,24 +288,54 @@ toMatrix m = do
...
+  return (CompatibleMatrices pn p s f)
 
   where
-    samePredNat :: (KnownNat m, KnownNat n)
-                => Proxy m -> Proxy n -> Maybe ((m - 1) :~: n)
-    samePredNat m n
-      | pred (natVal m) == natVal n = Just (unsafeCoerce Refl)
+    lemma0 :: (KnownNat m, KnownNat n) => Proxy m -> Proxy n -> Maybe (m :~: (n - 1))
+    lemma0 m n
+      | succ (natVal m) == natVal n = Just (unsafeCoerce Refl)
```
Comment (stevana) [inline src/Test/StateMachine/Markov.hs]: `natVal m == pred (natVal n)` makes more sense given the type?

File: src/Test/StateMachine/Markov.hs [hunk 4]
```diff
@@ -285,24 +288,54 @@ toMatrix m = do
...
       | otherwise                   = Nothing
+
+    lemma1 :: forall n. KnownNat (n - 1) => Proxy n -> Maybe (1 :<=? (n - 1))
+    lemma1 _ = pure $ (Proxy @1) %<=? (Proxy @(n - 1))
+
+    lemma2 :: Proxy n -> Maybe ((n - (n - 1)) :~: 1)
+    lemma2 _ = pure $ unsafeCoerce Refl
```
Comment (stevana) [inline src/Test/StateMachine/Markov.hs]: Use `sameNat`?

File: src/Test/StateMachine/Markov.hs [hunk 5]
```diff
@@ -285,24 +288,54 @@ toMatrix m = do
...
+
+    lemma3 :: forall n. (KnownNat n, KnownNat (n - 1)) => Proxy n -> Maybe ((n - 1) :<=? n)
+    lemma3 p' = pure $ (Proxy @(n - 1)) %<=? p'
+
+    lemma4 :: Proxy n -> Maybe ((n - n) :~: 0)
+    lemma4 _ = pure $ unsafeCoerce Refl
```
Comment (stevana) [inline src/Test/StateMachine/Markov.hs]: Use `sameNat`?

File: test/Spec.hs [hunk 6]
```diff
@@ -80,7 +81,7 @@ tests docker0 = testGroup "Tests"
           (expectFailure (ioProperty (prop_echoParallelOK True <$> mkEnv)))
       ]
   , testGroup "ProcessRegistry"
-      [ testProperty "sequential" (prop_processRegistry markovGood)
+      [ QCSM.testProperty markovGood "sequential" prop_processRegistry
```
Comment (stevana) [inline test/Spec.hs]: I think we should flip the arguments around here...

```haskell
QCSM.testProperty "sequential" prop_processRegistry markovGood
```
Is easier to read in my opinion.

File: src/Test/StateMachine/Markov.hs [hunk 7]
```diff
@@ -285,24 +288,54 @@ toMatrix m = do
 
 ------------------------------------------------------------------------
 
+instance IsString str => MonadFail (Either str) where
+  fail = Left . fromString
+
+handle :: Maybe a -> String -> Either String a
```
Comment (stevana) [inline src/Test/StateMachine/Markov.hs]: This combinator is called `maybeToRight` in https://hackage.haskell.org/package/either-5.0.1/docs/Data-Either-Combinators.html

File: src/Test/StateMachine/Markov.hs (hunk 2 above)
Comment (rdanitz) [reply src/Test/StateMachine/Markov.hs]: see https://ghc.haskell.org/trac/ghc/ticket/12160

**Pull request merged**